### ISSUE TITLE
Incorrect Implementation For secret_watcher.go

### ISSUE THREAD

**Issue by christmerlini** [projectcalico/confd#671]
Please forgive my format, I don't do this often.

According to https://github.com/projectcalico/calico/blob/master/security/comms/secure-bgp.md,
"Create (or update) a Kubernetes secret in the namespace where {{site.noderunning}} is running, so that it has a key whose value is the desired password. Note the secret name and the key name."

This means that confd should detect the namespace it it is currently running in to request the secret for BGP peering.

However according to this:
https://github.com/projectcalico/confd/blob/master/pkg/backends/calico/secret_watcher.go#L56-L61
It is assuming that the environment variable NAMESPACE is present with the current namespace name.

I have not been able to find documentation about this assertion.
According to Kubernetes: https://kubernetes.io/docs/tasks/run-application/access-api-from-pod/#directly-accessing-the-rest-api
Their documented method is to read /var/run/secrets/kubernetes.io/serviceaccount/namespace.

If specifying a custom namespace for the BGP secret is desired, I would recommend that a special purpose env be created like 'BGP_SECRET_NAMESPACE' and then if empty, default the namespace to whatever is in /var/run/secrets/kubernetes.io/serviceaccount/namespace.

If changing the current behavior becomes a breaking change, I would recommend that this hidden behavior be documented in https://github.com/projectcalico/calico/blob/master/security/comms/secure-bgp.md for clarification.

**Comment by caseydavenport** [comment]
I believe this env var is generally set by the operator when installing Calico: https://github.com/tigera/operator/blob/master/pkg/render/node.go#L1129-L1134

We don't really support specifying arbitrary namespaces to install Calico in - the manifests use kube-system (which the code defaults to) and the operator uses calico-system (and sets the env var appropriately). I don't think this env var was ever meant to be a user-specified config option, which is why it isn't currently documented.

**Comment by christmerlini** [comment]
@caseydavenport
I agree that using an arbitrary namespace is not officially supported by Calico.
However, I prefer not to install the tigera operator because it is too much of a blackbox for us and requires too much access to the cluster for what we need Calico for.
We also do not want to install Calico in the kube-system namespace with the non-operator way, to separate Calico from the core components, which makes creating RBACs easier.

Changing the code is probably just more work.
But at the same time, "Create (or update) a Kubernetes secret in the namespace where {{site.noderunning}} is running, so that it has a key whose value is the desired password. Note the secret name and the key name." from https://github.com/projectcalico/calico/blob/master/security/comms/secure-bgp.md is not what the code does either.

Will it be possible to have the following in the docs on https://github.com/projectcalico/calico/blob/master/security/comms/secure-bgp.md as a not under the first note regarding "BGP passwords must be 80 characters or fewer":

Note: To use a specific namespace to retrieve BGP secrets from, use the 'NAMESPACE' environment variable in the node container.

Or something like this that explains that you need use the 'NAMESPACE' env to run Calico in custom namespaces.

**Comment by caseydavenport** [comment]
> I agree that using an arbitrary namespace is not officially supported by Calico.

Unfortunately, adding documentation for this variable would effectively mean we need to change that position.

I agree it's unfortunate that the manifests run Calico in kube-system. I'd much rather figure out what concerns you have with the operator approach and address those than expose this variable. The operator is the future of how to install Calico on Kubernetes.

For now, I can't really stop you from using the NAMESPACE variable, but I don't want to document it as a supported configuration option. This is likely something that was missed in code review and shouldn't be there in the first place.
